huaweicloud/services/smn/resource_huaweicloud_smn_topic.go

```diff
@@ -208,6 +277,93 @@ func resourceTopicUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 	return resourceTopicRead(ctx, d, meta)
 }
 
+func updatePolicies(client *golangsdk.ServiceClient, d *schema.ResourceData, id string) error {
+	if d.HasChanges("users_publish_allowed", "services_publish_allowed") {
+		value, err := buildUpdateAccessPolicy(d)
+		if err != nil {
+			return nil
+		}
+		opts := topics.UpdatePoliciesOpts{
+			Value: value,
+		}
+		_, err = topics.UpdatePolicies(client, opts, id, "access_policy").Extract()
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("introduction") {
+		opts := topics.UpdatePoliciesOpts{
+			Value: d.Get("introduction").(string),
+		}
+		_, err := topics.UpdatePolicies(client, opts, id, "introduction").Extract()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func buildUpdateAccessPolicy(d *schema.ResourceData) (string, error) {
+	statement := []map[string]interface{}{}
+	if v, ok := d.GetOk("users_publish_allowed"); ok {
+		var csp interface{}
+		if v == "*" {
+			csp = "*"
+		} else {
+			csp = strings.Split(v.(string), ",")
+		}
+
+		statement = append(statement, map[string]interface{}{
+			"Sid":    "__user_pub_0",
+			"Effect": "Allow",
+			"Principal": map[string]interface{}{
+				"CSP": csp,
+			},
+			"Action": []interface{}{
+				"SMN:Publish",
+				"SMN:QueryTopicDetail",
+			},
+			"Resource": d.Id(),
+		})
+	}
+
+	if v, ok := d.GetOk("services_publish_allowed"); ok {
+		service := strings.Split(v.(string), ",")
+
+		statement = append(statement, map[string]interface{}{
+			"Sid":    "__service_pub_0",
+			"Effect": "Allow",
+			"Principal": map[string]interface{}{
+				"Service": service,
+			},
+			"Action": []interface{}{
+				"SMN:Publish",
+				"SMN:QueryTopicDetail",
+			},
+			"Resource": d.Id(),
+		})
+	}
+
+	if len(statement) == 0 {
```

Comment:
if both `users_publish_allowed` and `services_publish_allowed` are updated to empty, we will return empty and don‘t call the API